File: suc/swift/011.swift
import SwiftUI

@main
struct ForumPostApp: App {
    var body: some Scene {
        WindowGroup {
            ForumPostView()
                .statusBar(hidden: true)
        }
    }
}

struct Author {
    let name: String
    let avatarColor: Color
}

struct Reply: Identifiable {
    let id: Int
    let author: Author
    let timestamp: String
    let text: String
    let votes: Int
}

struct Post {
    let author: Author
    let timestamp: String
    let title: String
    let body: String
    let votes: Int
    let commentCount: Int
}

struct AppTokens {
    struct Colors {
        static let primary = Color(red: 0.0, green: 0.898, blue: 1.0)
        static let secondary = Color(red: 1.0, green: 0.0, blue: 0.667)
        static let tertiary = Color(red: 0.462, green: 1.0, blue: 0.012)
        static let background = Color(red: 0.063, green: 0.063, blue: 0.141)
        static let surface = Color(red: 0.11, green: 0.11, blue: 0.227)
        static let surfaceVariant = Color(red: 0.157, green: 0.157, blue: 0.314)
        static let outline = Color(red: 0.267, green: 0.267, blue: 0.416)
        static let warning = Color(red: 1.0, green: 0.839, blue: 0.0)
        static let onPrimary = Color.black
        static let onSurface = Color(red: 0.878, green: 0.878, blue: 1.0)
    }
    
    struct Typography {
        static let title = Font.system(size: 20, weight: .medium) // 增大
        static let headline = Font.system(size: 22, weight: .semibold) // 增大
        static let body = Font.system(size: 18, weight: .regular) // 增大
        static let label = Font.system(size: 16, weight: .medium) // 增大
        static let small = Font.system(size: 14, weight: .medium) // 增大
    }
    
    struct Shapes {
        static let small: CGFloat = 6
        static let medium: CGFloat = 12
        static let large: CGFloat = 20
    }
    
    struct Spacing {
        static let xs: CGFloat = 6
        static let sm: CGFloat = 10
        static let md: CGFloat = 16
        static let lg: CGFloat = 20
        static let xl: CGFloat = 28
    }
}

struct UpVoteArrow: View {
    let color: Color
    
    var body: some View {
        Canvas { context, size in
            let path = Path { path in
                path.move(to: CGPoint(x: size.width / 2, y: 0))
                path.addLine(to: CGPoint(x: size.width, y: size.height / 2))
                path.addLine(to: CGPoint(x: size.width / 2, y: size.height))
                path.addLine(to: CGPoint(x: 0, y: size.height / 2))
                path.closeSubpath()
            }
            context.fill(path, with: .color(color))
        }
        .frame(width: 28, height: 28) // 增大
    }
}

struct AuthorInfo: View {
    let author: Author
    let timestamp: String
    
    var body: some View {
        HStack(spacing: AppTokens.Spacing.sm) {
            ZStack {
                Circle()
                    .fill(author.avatarColor)
                    .frame(width: 28, height: 28) // 增大
                
                Circle()
                    .fill(Color.black.opacity(0.2))
                    .frame(width: 10, height: 10) // 增大
            }
            
            Text(author.name)
                .font(AppTokens.Typography.small)
                .fontWeight(.bold)
                .foregroundColor(author.avatarColor)
            
            Text(timestamp)
                .font(AppTokens.Typography.small)
                .foregroundColor(AppTokens.Colors.outline)
        }
    }
}

struct InfoChip: View {
    let text: String
    
    var body: some View {
        Text(text)
            .font(AppTokens.Typography.small)
            .fontWeight(.bold)
            .foregroundColor(AppTokens.Colors.outline)
    }
}

struct PostCard: View {
    let post: Post
    let voteState: Int
    let onVote: (Int) -> Void
    
    var body: some View {
        HStack(alignment: .top, spacing: AppTokens.Spacing.md) {
            VStack(spacing: AppTokens.Spacing.sm) {
                Button(action: {
                    onVote(voteState == 1 ? 0 : 1)
                }) {
                    UpVoteArrow(color: voteState == 1 ? AppTokens.Colors.secondary : AppTokens.Colors.outline)
                }
                
                Text("\(post.votes + voteState - 1)")
                    .font(AppTokens.Typography.title)
                    .fontWeight(.bold)
                    .foregroundColor(voteState != 0 ? AppTokens.Colors.secondary : AppTokens.Colors.onSurface)
                
                Button(action: {
                    onVote(voteState == -1 ? 0 : -1)
                }) {
                    UpVoteArrow(color: voteState == -1 ? AppTokens.Colors.primary : AppTokens.Colors.outline)
                        .rotationEffect(.degrees(180))
                }
            }
            .frame(width: 60) // 放宽投票区域
            
            VStack(alignment: .leading, spacing: AppTokens.Spacing.sm) {
                AuthorInfo(author: post.author, timestamp: post.timestamp)
                
                Text(post.title)
                    .font(AppTokens.Typography.headline)
                    .foregroundColor(AppTokens.Colors.onSurface)
                    .fixedSize(horizontal: false, vertical: true)
                
                Text(post.body)
                    .font(AppTokens.Typography.body)
                    .foregroundColor(AppTokens.Colors.onSurface.opacity(0.8))
                    .fixedSize(horizontal: false, vertical: true)
                
                HStack(spacing: AppTokens.Spacing.xl) {
                    InfoChip(text: "\(post.commentCount) Comments")
                    InfoChip(text: "Share")
                }
                .padding(.top, AppTokens.Spacing.sm)
            }
        }
        .padding(AppTokens.Spacing.lg)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(AppTokens.Colors.surface)
    }
}

struct ReplyCard: View {
    let reply: Reply
    
    var body: some View {
        HStack(alignment: .top, spacing: AppTokens.Spacing.md) {
            Rectangle()
                .fill(reply.author.avatarColor.opacity(0.5))
                .frame(width: 3) // 增大
                .cornerRadius(1.5)
            
            VStack(alignment: .leading, spacing: AppTokens.Spacing.xs) {
                AuthorInfo(author: reply.author, timestamp: reply.timestamp)
                
                Text(reply.text)
                    .font(AppTokens.Typography.body)
                    .foregroundColor(AppTokens.Colors.onSurface.opacity(0.9))
                    .fixedSize(horizontal: false, vertical: true)
                
                HStack(spacing: AppTokens.Spacing.md) {
                    InfoChip(text: "\(reply.votes) votes")
                    InfoChip(text: "Reply")
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }
}

struct TopAppBar: View {
    var body: some View {
        ZStack {
            Rectangle()
                .fill(AppTokens.Colors.surface)
                .frame(height: 60) // 增大
                .ignoresSafeArea(edges: .top)
            
            HStack {
                Button(action: {}) {
                    ZStack {
                        Canvas { context, size in
                            let centerY = size.height / 2
                            
                            context.stroke(
                                Path { path in
                                    path.move(to: CGPoint(x: 0, y: centerY))
                                    path.addLine(to: CGPoint(x: size.width, y: centerY))
                                },
                                with: .color(AppTokens.Colors.primary),
                                lineWidth: 3 // 增大
                            )
                            
                            context.stroke(
                                Path { path in
                                    path.move(to: CGPoint(x: 0, y: centerY))
                                    path.addLine(to: CGPoint(x: size.width / 2, y: 0))
                                },
                                with: .color(AppTokens.Colors.primary),
                                lineWidth: 3 // 增大
                            )
                            
                            context.stroke(
                                Path { path in
                                    path.move(to: CGPoint(x: 0, y: centerY))
                                    path.addLine(to: CGPoint(x: size.width / 2, y: size.height))
                                },
                                with: .color(AppTokens.Colors.primary),
                                lineWidth: 3 // 增大
                            )
                        }
                        .frame(width: 24, height: 24) // 增大
                    }
                    .frame(width: 50, height: 50) // 增大
                }
                
                Spacer()
                
                Text("r/jetpackcompose")
                    .font(AppTokens.Typography.title)
                    .foregroundColor(AppTokens.Colors.onSurface)
                
                Spacer()
                
                Rectangle()
                    .fill(Color.clear)
                    .frame(width: 50, height: 50) // 增大
            }
            .padding(.horizontal, AppTokens.Spacing.lg)
        }
    }
}

struct BottomCommentBar: View {
    @Binding var commentText: String
    
    var body: some View {
        VStack(spacing: 0) {
            Rectangle()
                .fill(AppTokens.Colors.outline.opacity(0.3))
                .frame(height: 1) // 增大
                .padding(.horizontal, AppTokens.Spacing.lg)
            
            HStack(spacing: AppTokens.Spacing.md) {
                ZStack(alignment: .leading) {
                    RoundedRectangle(cornerRadius: AppTokens.Shapes.large)
                        .fill(AppTokens.Colors.surfaceVariant)
                        .frame(height: 52) // 增大
                    
                    if commentText.isEmpty {
                        Text("Add a comment...")
                            .font(AppTokens.Typography.body)
                            .foregroundColor(AppTokens.Colors.outline)
                            .padding(.horizontal, AppTokens.Spacing.lg)
                    }
                    
                    TextField("", text: $commentText)
                        .font(AppTokens.Typography.body)
                        .foregroundColor(AppTokens.Colors.onSurface)
                        .padding(.horizontal, AppTokens.Spacing.lg)
                }
                
                Button(action: {}) {
                    Text("POST")
                        .font(AppTokens.Typography.label)
                        .fontWeight(.bold)
                        .foregroundColor(AppTokens.Colors.onPrimary)
                        .padding(.horizontal, AppTokens.Spacing.lg)
                        .frame(height: 52) // 增大
                        .background(
                            RoundedRectangle(cornerRadius: AppTokens.Shapes.large)
                                .fill(AppTokens.Colors.primary)
                        )
                }
            }
            .padding(.horizontal, AppTokens.Spacing.lg)
            .padding(.vertical, AppTokens.Spacing.md)
            .background(AppTokens.Colors.surface)
        }
    }
}

struct ForumPostView: View {
    @State private var voteState = 1
    @State private var commentText = ""
    
    private let post = Post(
        author: Author(name: "user/compose_master", avatarColor: AppTokens.Colors.tertiary),
        timestamp: "8 hours ago",
        title: "The power of Jetpack Compose for single-file UIs",
        body: "Creating fully functional, beautiful UIs in a single Kotlin file is not just possible, it's becoming a new standard for prototypes and minimalist apps. What are your thoughts on this approach?",
        votes: 1337,
        commentCount: 42
    )
    
    private let replies = [
        Reply(id: 1, author: Author(name: "dev/android_fan", avatarColor: AppTokens.Colors.secondary), timestamp: "7h ago", text: "Absolutely agree! It simplifies dependency management and makes sharing snippets a breeze.", votes: 25),
        Reply(id: 2, author: Author(name: "ux/designer_jane", avatarColor: AppTokens.Colors.primary), timestamp: "5h ago", text: "From a design perspective, having the design tokens and theme right next to the UI components is a game-changer for consistency.", votes: 18)
    ]
    
    var body: some View {
        VStack(spacing: 0) {
            TopAppBar()
            
            ScrollView {
                LazyVStack(spacing: 0) {
                    PostCard(post: post, voteState: voteState, onVote: { newVote in
                        voteState = newVote
                    })
                    
                    Text("TOP COMMENTS")
                        .font(AppTokens.Typography.small)
                        .foregroundColor(AppTokens.Colors.outline)
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .padding(.horizontal, AppTokens.Spacing.lg)
                        .padding(.vertical, AppTokens.Spacing.sm)
                    
                    ForEach(replies) { reply in
                        ReplyCard(reply: reply)
                            .padding(.horizontal, AppTokens.Spacing.lg)
                            .padding(.vertical, AppTokens.Spacing.sm)
                    }
                }
            }
            .background(AppTokens.Colors.background)
            
            BottomCommentBar(commentText: $commentText)
        }
        .background(AppTokens.Colors.background)
    }
}

File: suc/kotlin/011.kt
package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CenterAlignedTopAppBar
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Shapes
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.material3.Typography
import androidx.compose.material3.darkColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.painter.Painter
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.graphics.vector.rememberVectorPainter
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat

private const val NAME = "008_ForumPost_en"
private const val UI_TYPE = "Social Forum Post"
private const val STYLE_THEME = "Dark Neon"
private const val LANG = "en"
private const val BASELINE_SIZE = "720x1280"

object AppTokens {
    object Colors {
        val primary = Color(0xFF00E5FF)
        val secondary = Color(0xFFFF00AA)
        val tertiary = Color(0xFF76FF03)
        val background = Color(0xFF101024)
        val surface = Color(0xFF1C1C3A)
        val surfaceVariant = Color(0xFF282850)
        val outline = Color(0xFF44446A)
        val success = Color(0xFF00C853)
        val warning = Color(0xFFFFD600)
        val error = Color(0xFFFF1744)
        val onPrimary = Color(0xFF000000)
        val onSecondary = Color(0xFF000000)
        val onTertiary = Color(0xFF000000)
        val onBackground = Color(0xFFE0E0FF)
        val onSurface = Color(0xFFE0E0FF)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 28.sp, fontWeight = FontWeight.Bold, letterSpacing = 0.sp)
        val headline = TextStyle(fontSize = 22.sp, fontWeight = FontWeight.SemiBold, letterSpacing = 0.sp)
        val title = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.Medium, letterSpacing = 0.sp)
        val body = TextStyle(fontSize = 16.sp, fontWeight = FontWeight.Normal, lineHeight = 24.sp, letterSpacing = 0.sp)
        val label = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.Medium, letterSpacing = 0.5.sp)
    }
    object Shapes {
        val small = RoundedCornerShape(4.dp)
        val medium = RoundedCornerShape(8.dp)
        val large = RoundedCornerShape(16.dp)
    }
    object Spacing {
        val xs = 4.dp
        val sm = 8.dp
        val md = 12.dp
        val lg = 16.dp
        val xl = 24.dp
        val xxl = 32.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(1.dp, 2.dp, 1.dp, 0.1f)
        val level2 = ShadowSpec(3.dp, 4.dp, 2.dp, 0.12f)
        val level3 = ShadowSpec(6.dp, 8.dp, 4.dp, 0.14f)
    }
}

private val AppColorScheme = darkColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    onSecondary = AppTokens.Colors.onSecondary,
    tertiary = AppTokens.Colors.tertiary,
    onTertiary = AppTokens.Colors.onTertiary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onBackground,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface,
    surfaceVariant = AppTokens.Colors.surfaceVariant,
    outline = AppTokens.Colors.outline,
    error = AppTokens.Colors.error
)

private val AppTypography = Typography(
    displaySmall = AppTokens.TypographyTokens.display,
    headlineSmall = AppTokens.TypographyTokens.headline,
    titleMedium = AppTokens.TypographyTokens.title,
    bodyLarge = AppTokens.TypographyTokens.body,
    labelMedium = AppTokens.TypographyTokens.label
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        shapes = Shapes(
            small = AppTokens.Shapes.small,
            medium = AppTokens.Shapes.medium,
            large = AppTokens.Shapes.large
        ),
        content = content
    )
}

data class Author(val name: String, val avatarColor: Color)
data class Reply(val id: Int, val author: Author, val timestamp: String, val text: String, val votes: Int)
data class Post(val author: Author, val timestamp: String, val title: String, val body: String, val votes: Int, val commentCount: Int)

@Composable
fun UpVoteArrow(modifier: Modifier = Modifier, color: Color) {
    Canvas(modifier = modifier.size(24.dp)) {
        val path = Path().apply {
            moveTo(size.width / 2, 0f)
            lineTo(size.width, size.height / 2)
            lineTo(size.width / 2, size.height)
            lineTo(0f, size.height / 2)
            close()
        }
        drawPath(path, color)
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RootScreen() {
    val post = Post(
        author = Author("user/compose_master", AppTokens.Colors.tertiary),
        timestamp = "8 hours ago",
        title = "The power of Jetpack Compose for single-file UIs",
        body = "Creating fully functional, beautiful UIs in a single Kotlin file is not just possible, it\'s becoming a new standard for prototypes and minimalist apps. What are your thoughts on this approach?",
        votes = 1337,
        commentCount = 42
    )
    val replies = remember {
        listOf(
            Reply(1, Author("dev/android_fan", AppTokens.Colors.secondary), "7h ago", "Absolutely agree! It simplifies dependency management and makes sharing snippets a breeze.", 25),
            Reply(2, Author("ux/designer_jane", AppTokens.Colors.primary), "5h ago", "From a design perspective, having the design tokens and theme right next to the UI components is a game-changer for consistency.", 18),
            Reply(3, Author("newbie/coder", AppTokens.Colors.warning), "2h ago", "I found it a bit overwhelming at first. Any tips for organizing the file?", 3)
        )
    }
    var voteState by remember { mutableStateOf(1) }

    Scaffold(
        contentWindowInsets = androidx.compose.foundation.layout.WindowInsets(0),
        containerColor = AppTokens.Colors.background,
        topBar = {
            CenterAlignedTopAppBar(
                title = { Text("r/jetpackcompose", style = MaterialTheme.typography.titleMedium, color = MaterialTheme.colorScheme.onSurface) },
                navigationIcon = {
                    IconButton(onClick = {}) {
                        Canvas(modifier = Modifier.size(20.dp)) {
                            drawLine(color = AppTokens.Colors.primary, start = Offset(0f, center.y), end = Offset(size.width, center.y), strokeWidth = 2.dp.toPx())
                            drawLine(color = AppTokens.Colors.primary, start = Offset(0f, center.y), end = Offset(size.width/2, 0f), strokeWidth = 2.dp.toPx())
                            drawLine(color = AppTokens.Colors.primary, start = Offset(0f, center.y), end = Offset(size.width/2, size.height), strokeWidth = 2.dp.toPx())
                        }
                    }
                },
                colors = TopAppBarDefaults.centerAlignedTopAppBarColors(containerColor = AppTokens.Colors.surface)
            )
        },
        bottomBar = {
            Surface(color = AppTokens.Colors.surface, tonalElevation = AppTokens.ElevationMapping.level3.elevation) {
                Row(
                    modifier = Modifier.fillMaxWidth().padding(AppTokens.Spacing.lg, AppTokens.Spacing.md),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md)
                ) {
                    Box(modifier = Modifier.weight(1f).height(48.dp).clip(AppTokens.Shapes.large).background(AppTokens.Colors.surfaceVariant).padding(horizontal = AppTokens.Spacing.lg), contentAlignment = Alignment.CenterStart) {
                        Text("Add a comment...", color = AppTokens.Colors.outline, style = MaterialTheme.typography.bodyLarge)
                    }
                    Button(onClick = {}, shape = AppTokens.Shapes.large, colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.primary), contentPadding = PaddingValues(AppTokens.Spacing.lg)) {
                        Text("POST", style = MaterialTheme.typography.labelMedium, color = AppTokens.Colors.onPrimary, fontWeight = FontWeight.Bold)
                    }
                }
            }
        }
    ) { paddingValues ->
        LazyColumn(
            modifier = Modifier.fillMaxSize().padding(paddingValues),
            contentPadding = PaddingValues(vertical = AppTokens.Spacing.lg)
        ) {
            item { PostCard(post = post, voteState = voteState, onVote = { voteState = it }) }
            item {
                Text(
                    "TOP COMMENTS",
                    style = MaterialTheme.typography.labelMedium,
                    color = AppTokens.Colors.outline,
                    modifier = Modifier.padding(horizontal = AppTokens.Spacing.lg, vertical = AppTokens.Spacing.sm)
                )
            }
            items(replies) { reply ->
                ReplyCard(reply = reply, modifier = Modifier.padding(horizontal = AppTokens.Spacing.lg, vertical = AppTokens.Spacing.sm))
            }
        }
    }
}

@Composable
fun PostCard(post: Post, voteState: Int, onVote: (Int) -> Unit) {
    Row(
        modifier = Modifier.fillMaxWidth().background(AppTokens.Colors.surface).padding(AppTokens.Spacing.lg),
        horizontalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md)
    ) {
        Column(horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.sm)) {
            IconButton(onClick = { onVote(if (voteState == 1) 0 else 1) }) {
                UpVoteArrow(color = if (voteState == 1) AppTokens.Colors.secondary else AppTokens.Colors.outline)
            }
            Text((post.votes + voteState - 1).toString(), style = MaterialTheme.typography.titleMedium, color = if (voteState != 0) AppTokens.Colors.secondary else AppTokens.Colors.onSurface, fontWeight = FontWeight.Bold)
            IconButton(onClick = { onVote(if (voteState == -1) 0 else -1) }) {
                Box(modifier = Modifier.size(24.dp)) {
                    UpVoteArrow(color = if (voteState == -1) AppTokens.Colors.primary else AppTokens.Colors.outline, modifier = Modifier.matchParentSize().align(Alignment.Center))
                }
            }
        }
        Column(verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.sm)) {
            AuthorInfo(post.author, post.timestamp)
            Text(post.title, style = MaterialTheme.typography.headlineSmall, color = AppTokens.Colors.onSurface)
            Text(post.body, style = MaterialTheme.typography.bodyLarge, color = AppTokens.Colors.onSurface.copy(alpha = 0.8f))
            Spacer(Modifier.height(AppTokens.Spacing.sm))
            Row(horizontalArrangement = Arrangement.spacedBy(AppTokens.Spacing.xl)) {
                InfoChip(text = "${post.commentCount} Comments")
                InfoChip(text = "Share")
            }
        }
    }
}

@Composable
fun ReplyCard(reply: Reply, modifier: Modifier = Modifier) {
    Row(
        modifier = modifier.fillMaxWidth().height(intrinsicSize = IntrinsicSize.Min),
        horizontalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md)
    ) {
        Box(
            modifier = Modifier.fillMaxHeight().width(2.dp).background(
                color = reply.author.avatarColor.copy(alpha = 0.5f),
                shape = RoundedCornerShape(2.dp)
            )
        )
        Column(verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.xs)) {
            AuthorInfo(reply.author, reply.timestamp)
            Text(reply.text, style = MaterialTheme.typography.bodyLarge, color = AppTokens.Colors.onSurface.copy(alpha = 0.9f))
            Row(horizontalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md), verticalAlignment = Alignment.CenterVertically) {
                InfoChip(text = "${reply.votes} votes")
                InfoChip(text = "Reply")
            }
        }
    }
}

@Composable
fun AuthorInfo(author: Author, timestamp: String) {
    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(AppTokens.Spacing.sm)) {
        Box(modifier = Modifier.size(24.dp).clip(CircleShape).background(author.avatarColor)) {
            Canvas(modifier = Modifier.matchParentSize()) {
                drawCircle(color = Color.Black.copy(alpha=0.2f), radius = size.minDimension/3)
            }
        }
        Text(author.name, style = MaterialTheme.typography.labelMedium, color = author.avatarColor, fontWeight = FontWeight.Bold)
        Text(timestamp, style = MaterialTheme.typography.labelMedium, color = AppTokens.Colors.outline)
    }
}

@Composable
fun InfoChip(text: String) {
    Text(text, style = MaterialTheme.typography.labelMedium, color = AppTokens.Colors.outline, fontWeight = FontWeight.Bold)
}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.systemBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.systemBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFF101024)
@Composable
fun PreviewRoot() {
    AppTheme {
        RootScreen()
    }
}